base/abstractarray.jl

```diff
@@ -290,15 +297,22 @@ end
 function copy!(dest::AbstractArray, doffs::Integer, src::AbstractArray)
     copy!(dest, doffs, src, 1, length(src))
 end
+
 function copy!(dest::AbstractArray, doffs::Integer, src::AbstractArray, soffs::Integer)
-    soffs > length(src) && throw(BoundsError())
+    soffs > length(src) && throw(BoundsError(src, soffs))
     copy!(dest, doffs, src, soffs, length(src)-soffs+1)
 end
-function copy!(dest::AbstractArray, doffs::Integer, src::AbstractArray, soffs::Integer, n::Integer)
-    n < 0 && throw(BoundsError())
+
+function copy!(dest::AbstractArray, doffs::Integer,
+               src::AbstractArray, soffs::Integer,
+               n::Integer)
+    n < 0 && throw(BoundsError(src, n))
     n == 0 && return dest
     if soffs+n-1 > length(src) || doffs+n-1 > length(dest) || doffs < 1 || soffs < 1
-        throw(BoundsError())
+        soffs+n-1 > length(src)  && throw(BoundsError(src, soffs+n-1))
+        doffs+n-1 > length(dest) && throw(BoundsError(dest, doffs+n-1))
+        doffs < 1 && throw(BoundsError(dest, doffs))
+        throw(BoundsError(src, soffs))
     end
```

Comment:
In this case, these were already short-circuiting comparisons.  We may as well remove the outer if statement.